llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp

```diff
@@ -110,17 +110,25 @@ StringRef NVPTXLowerUnreachable::getPassName() const {
 }
 
 // =============================================================================
-// Returns whether a `trap` intrinsic should be emitted before I.
+// Returns whether a `trap` intrinsic would be emitted before I.
 //
 // This is a copy of the logic in SelectionDAGBuilder::visitUnreachable().
 // =============================================================================
 bool NVPTXLowerUnreachable::isLoweredToTrap(const UnreachableInst &I) const {
-  if (!TrapUnreachable)
-    return false;
-  if (!NoTrapAfterNoreturn)
-    return true;
   const CallInst *Call = dyn_cast_or_null<CallInst>(I.getPrevNode());
-  return Call && Call->doesNotReturn();
+  if (Call) {
+    // We've already emitted a non-continuable trap.
+    if (Call->isNonContinuableTrap())
+      return true;
```

Comment:
> this pass ... would generate traps even if the instruction directly before the UnreachableInst was llvm.trap()

Why is this a problem?

Reply:
It's not in the sense that there's no miscompile involved, but correctly matching the logic in the instruction selectors happens to change this case too.

Reply:
Can you provide a link to where this logic malfunctions in the instruction selector?

Reply:
The instruction selector is correct -- logic is at `SelectionDAGBuilder::visitUnreachable`